kernel/mmanutils.c

```diff
@@ -517,219 +571,288 @@ int myst_get_free_ram(size_t* size)
         ERAISE(-ENOMEM);
 
     myst_buf_clear(vbuf);
 
-    myst_spin_lock(&_mappings_lock);
     {
-        for (myst_process_mapping_t* p = _mappings; p; p = p->next)
+        uint8_t* addr = (uint8_t*)_mman.map;
+        size_t length = ((uint8_t*)_mman.end) - addr;
+        size_t index;
+        vectors_t v = _get_vectors();
+
+        assert(v.pids_count == v.fdmappings_count);
+
+        ECHECK(myst_round_up(length, PAGE_SIZE, &length));
+        ECHECK((index = _get_page_index(addr, length)));
+        size_t count = length / PAGE_SIZE;
+
+        assert(index < v.pids_count);
+        assert(index + count <= v.pids_count);
+
+        myst_rspin_lock(&_mman.lock);
         {
-            if (p->pid == pid && p->pathname)
+            for (size_t i = index; i < index + count;)
             {
-                // ATTN: device and inode number are not reported.
-                // Shared or private perms bit is always marked 'p',
-                // as MAP_SHARED is not supported.
-                int n = snprintf(
-                    locals->maps_entry,
-                    sizeof(locals->maps_entry),
-                    "%08lx-%08lx %c%c%cp %08lx 00:00 0 %s\n",
-                    (long)p->addr,
-                    (long)p->addr + p->size,
-                    p->prot & PROT_READ ? 'r' : '-',
-                    p->prot & PROT_WRITE ? 'w' : '-',
-                    p->prot & PROT_EXEC ? 'x' : '-',
-                    p->offset,
-                    p->pathname);
-
-                if (n >= (int)sizeof(locals->maps_entry))
-                    ERAISE(-ERANGE);
-
-                // Insert new entry at beginning
-                ECHECK(myst_buf_insert(
-                    vbuf, 0, locals->maps_entry, strlen(locals->maps_entry)));
+                if (v.pids[i] == (uint32_t)pid)
+                {
+                    size_t n = 1;
+                    size_t len;
+                    int fd = v.fdmappings[i].fd;
+
+                    uint32_t used = v.fdmappings[i].used;
+                    uint64_t offset = v.fdmappings[i].offset;
+                    myst_refstr_t* pathname = v.fdmappings[i].pathname;
+                    int prot = 0;
+                    bool consistent = false;
+                    char* str;
+                    int flags = 0;
+
+                    if (myst_mman_get_prot(
+                            &_mman, addr, PAGE_SIZE, &prot, &consistent) != 0)
+                    {
+                        assert("myst_mman_get_prot() failed\n");
+                    }
+
+                    /* count consecutive pages with same traits */
+                    for (size_t j = i + 1; j < index + count; j++)
+                    {
+                        int tmp_prot = 0;
+
+                        /* if the pid changes */
+                        if (v.pids[j] != (uint32_t)pid)
+                            break;
+
+                        if (v.fdmappings[i].used != used)
```

Comment:
Should the index variable be `j`, as the others?

Reply:
Oh my! Great catch! Fixed and running pipeline again.